juju/juju
# relocated AddCharm and AddCharmWithAuth to charms facade and use.

## Description of change

This work is in preparation for deploying a CharmHub charm.  AddCharm and AddCharmWithAuthorization facades calls have been updated to both take a charm origin and return one.  This is needed for CharmHub charms to provide an ID found when resolving the charm so it can be used when downloading the charm for deploy or upgrade.  Then passed passed back during the Deploy call and saved.  The ID helps tracking on the charm hub side.

Currently AddCharm and AddCharmWithAuthorization are done via the api client, which then calls a function on the application facade.  Neither of these are versioned.  Deprecate those functions and move the charms facade where they will be versioned.

## QA steps

There should be no change in current functionality deploying local and charm store charms.  Please try each of the following:

1. deploy a cs charm
2. deploy a cs charm at an older revision then upgrade it.
3. deploy a local charm
3. deploy a bundle.
